esi/tests/unit/lib/test_nodes.py

```diff
@@ -0,0 +1,505 @@
+        })
+        self.port2 = utils.create_mock_object({
+            "uuid": "port_uuid_2",
+            "node_id": "11111111-2222-3333-4444-bbbbbbbbbbbb",
+            "address": "bb:bb:bb:bb:bb:bb",
+            "internal_info": {}
+        })
+        self.port3 = utils.create_mock_object({
+            "uuid": "port_uuid_3",
+            "node_id": "11111111-2222-3333-4444-bbbbbbbbbbbb",
+            "address": "cc:cc:cc:cc:cc:cc",
+            "internal_info": {'tenant_vif_port_id': 'neutron_port_uuid_2'}
+        })
+        self.port4 = utils.create_mock_object({
+            "uuid": "port_uuid_4",
+            "node_id": "11111111-2222-3333-4444-bbbbbbbbbbbb",
+            "address": "dd:dd:dd:dd:dd:dd",
+            "internal_info": {'tenant_vif_port_id': 'neutron_port_uuid_4'}
+        })
+        self.node1 = utils.create_mock_object({
+            "id": "11111111-2222-3333-4444-aaaaaaaaaaaa",
+            "name": "node1"
+        })
+        self.node2 = utils.create_mock_object({
+            "id": "11111111-2222-3333-4444-bbbbbbbbbbbb",
+            "name": "node2"
+        })
+
+        self.connection = mock.Mock()
+
+        def mock_find_node(name_or_id=None, ignore_missing=True):
+            if name_or_id == "11111111-2222-3333-4444-aaaaaaaaaaaa" or name_or_id == "node1":
+                return self.node1
+            elif name_or_id == "11111111-2222-3333-4444-bbbbbbbbbbbb" or name_or_id == "node2":
+                return self.node2
+        self.connection.baremetal.find_node.side_effect = mock_find_node
+
+        def mock_baremetal_ports(details=False, node_id=None):
+            if node_id == '11111111-2222-3333-4444-aaaaaaaaaaaa':
+                return [self.port1]
+            if node_id == '11111111-2222-3333-4444-bbbbbbbbbbbb':
+                return [self.port2, self.port3, self.port4]
+            return [self.port1, self.port2, self.port3, self.port4]
+        self.connection.baremetal.ports.side_effect = mock_baremetal_ports
+
+    def test_node_and_port_list(self):
+        self.connection.baremetal.nodes.\
+            return_value = [self.node1, self.node2]
+
+        actual = nodes.node_and_port_list(self.connection, filter_node=None)
+        expected = (
+            [
+                self.node1,
+                self.node2
+            ],
+            [
+                self.port1,
+                self.port2,
+                self.port3,
+                self.port4,
+            ]
+        )
+        self.assertEqual(expected, actual)
+
+    def test_node_and_port_list_node_id_filter(self):
+        actual = nodes.node_and_port_list(self.connection, filter_node=self.node1.id)
+        expected = (
+            [
+                self.node1,
+            ],
+            [
+                self.port1,
+            ]
+        )
+        self.assertEqual(actual, expected)
+
+    def test_node_and_port_list_node_name_filter(self):
```

Comment:
You don't actually need this test case, since `filter_node` is treated the same in the function regardless of whether it's actually and ID or a name.